be/src/storage/lake/pk_index_loader.cpp

```diff
@@ -0,0 +1,181 @@
+            // more than one key column
+            if (!PrimaryKeyEncoder::create_column(_schema, &pk_column).ok()) {
+                CHECK(false) << "create column for primary key encoder failed";
+            }
+        }
+        while (true) {
+            chunk->reset();
+            rowids.clear();
+            st = (*seg_itr)->get_next(chunk, &rowids);
+            if (st.is_end_of_file()) {
+                break;
+            } else if (!st.ok()) {
+                return;
+            } else {
+                Column* pkc = nullptr;
+                if (pk_column) {
+                    pk_column->reset_column();
+                    PrimaryKeyEncoder::encode(_schema, *chunk, 0, chunk->num_rows(), pk_column.get());
+                    pkc = pk_column.get();
+                } else {
+                    pkc = chunk->columns()[0].get();
+                }
+                st = _index->insert(_rowset_id + _seg_id, rowids, *pkc);
+                if (!st.ok()) {
+                    return;
+                }
+            }
+        }
+        (*seg_itr)->close();
+    }
+
+private:
+    Tablet* _tablet;
+    uint32_t _rowset_id;
+    uint32_t _seg_id;
+    std::string _seg_name;
+    const Schema _schema;
+    int64_t _version;
+    const MetaFileBuilder* _builder;
+    LakePrimaryIndex* _index;
+};
+
+Status PkIndexLoader::init() {
+    return ThreadPoolBuilder("lake_load_pk_index")
+            .set_max_threads(config::lake_pk_index_loader_thread)
+            .build(&_load_thread_pool);
+}
+
+PkIndexLoader::~PkIndexLoader() {
+    if (_load_thread_pool) {
+        _load_thread_pool->shutdown();
+    }
+}
+
+std::future<Status> PkIndexLoader::load(Tablet* tablet, const std::vector<RowsetPtr>& rowsets, const Schema& schema,
+                                        int64_t version, const MetaFileBuilder* builder, LakePrimaryIndex* index) {
+    uint64_t subtask_num = 0;
+    auto p = std::make_unique<std::promise<Status>>();
+    std::future<Status> f = p->get_future();
+    std::vector<std::shared_ptr<Runnable>> subtasks;
+    for (auto& rowset : rowsets) {
+        auto& rowset_meta = rowset->metadata();
+        for (uint32_t i = 0; i < rowset_meta.segments_size(); ++i) {
+            std::shared_ptr<Runnable> subtask(std::make_shared<LoadPkIndexSubtask>(
+                    tablet, rowset->id(), i, rowset_meta.segments(i), schema, version, builder, index));
+            subtasks.push_back(subtask);
+        }
+        subtask_num += rowset_meta.segments_size();
+    }
+    if (subtask_num == 0) {
+        p->set_value(Status::OK());
+        return f;
+    }
+
+    std::lock_guard<std::mutex> lg(_mutex);
+    for (auto& subtask : subtasks) {
+        auto st = _load_thread_pool->submit(std::move(subtask));
+        if (!st.ok()) {
+            LOG(ERROR) << "Fail to submit pk index loading subtask, error: " << st.to_string();
+            subtask_num--;
+        }
+    }
+    _subtask_nums.emplace(tablet->id(), subtask_num);
```

Comment:
I think use LatchDown is more simple here, instead of `_mutex`, `_subtask_nums` and `_promises`.

Reply:
> I think use LatchCountDown is more simple here, instead of `_mutex`, `_subtask_nums` and `_promises`.

But we need to return status while `CountDownLatch` can't.